### ISSUE TITLE
`item.ident` for `ItemKind::Impl`

### ISSUE THREAD

**Issue by bruno-medeiros** [serde-rs/syntex#106]
It used to be with syntax_syntex version 0.28 or that for `item: &Item` that were of kind `ItemKind::Impl` , the `item.ident` would have the identifier that the impl applied to. However, in recent versions of syntax_syntex, that no longer seems to be the case, the ident is just the empty string. Is there still some straightforward way to obtain that information?

The ident for impls was used by the [Rainicorn tool](https://github.com/RustDT/Rainicorn) to provide Outline/Structure view for IDES.

**Comment by dtolnay** [comment]
This isn't up to Syntex, we just copy the parser out of the Rust compiler. I tracked this down to the following pull request from March: [rust-lang/rust#32570](https://github.com/rust-lang/rust/pull/32570/commits/7bebe80bc28ff818c7990061275fa0ad6c8b3686#diff-da9d34ca1d0f6beee2838cf02e07345cR5123). @eddyb can I interest you in reintroducing `impl_pretty_name`?

Failing that, you can always build the ident in Rainicorn. The `ItemKind::Impl` variant looks like:

```rust
Impl(Unsafety, ImplPolarity, Generics, Option<TraitRef>, P<Ty>, Vec<ImplItem>)
```

The `Option<TraitRef>` and `P<Ty>` were used to build the name. The function was:

```rust
pub fn impl_pretty_name(trait_ref: &Option<TraitRef>, ty: Option<&Ty>) -> Ident {
    let mut pretty = match ty {
        Some(t) => pprust::ty_to_string(t),
        None => String::from("..")
    };

    match *trait_ref {
        Some(ref trait_ref) => {
            pretty.push('.');
            pretty.push_str(&pprust::path_to_string(&trait_ref.path));
        }
        None => {}
    }
    token::gensym_ident(&pretty[..])
}
```

**Comment by eddyb** [comment]
@dtolnay No, it was a horrible hack. Usually if you want that kind of information, you have a `tcx: TyCtxt` and can use `tcx.item_path_str`, which does it in a more principled way (assuming you can't do better).

**Comment by bruno-medeiros** [comment]
What's a `TyCtxt` and how do I get one? 

I should note that I pretty much just want the identifier from an `impl` definition and not do any semantic analysis or anything. That is, for `impl Foo { }`, I don't even need a type `Foo` to be defined anywhere, since `Foo` will just be used for display purposes in my tool.

**Comment by erickt** [comment]
@eddyb: this is in syntex, so we don't have a `TyCtxt`. Is there another way we can get this purely through libsyntax?

**Comment by eddyb** [comment]
Not really, no, you can't get anything semantically interesting in that case, so the hack is one of the better options.

**Comment by bruno-medeiros** [comment]
> Not really, no, you can't get anything semantically interesting in that case, so the hack is one of the better options.

Like I said, I don't want anything semantic for this, just pure syntax. I will give @dtolnay code a try then, and report back 👍

**Comment by bruno-medeiros** [comment]
> Not really, no, you can't get anything semantically interesting in that case, so the hack is one of the better options.

Like I said,  for this use case I don't want anything semantic, just pure syntax. I will give @dtolnay code a try then, and report back 👍
